external/fv3fit/fv3fit/pytorch/graph/network.py

```diff
@@ -0,0 +1,111 @@
+import dataclasses
+from typing import Callable
+import torch.nn.functional as F
+import torch
+import torch.nn as nn
+from dgl.nn.pytorch import SAGEConv
+from .graph_builder import build_dgl_graph
+
+
+@dataclasses.dataclass
+class GraphNetworkConfig:
+    """
+    Attributes:
+        n_hidden: maximum number of hidden channels
+        num_blocks: number of nested network (e.g, 2 means N(N(theta)))
+        aggregator: type of aggregator, one of "mean", "gcn", "pool", or "lstm"
+        activation: activation function
+    """
+
+    n_hidden: int = 256
+    num_blocks: int = 1
+    aggregator: str = "mean"
+    activation: Callable = F.relu
+
+
+class CubedSphereGraphOperation(nn.Module):
+    """
+    A wrapper class which applies graph operations to cubed sphere data.
+    """
+
+    def __init__(self, graph_op: nn.Module):
+        """
+        Initializes the wrapped graph operation.
+
+        Args:
+            graph_op: graph operation to apply, whose first argument should be
+                a DGL graph and second argument should be a tensor of shape
+                (n_tiles * n_x * n_y, batch_size, n_features)
+        """
+        super().__init__()
+        self.graph_op = graph_op
+
+    def forward(self, inputs):
+        """
+        Args:
+            inputs: tensor of shape (batch_size, n_tiles, n_x, n_y, n_features)
+
+        Returns:
+            tensor of shape (batch_size, n_tiles, n_x, n_y, n_features_out)
+        """
+        if len(inputs.shape) != 5:
+            raise ValueError(
+                "inputs must be of shape (batch_size, n_tiles, n_x, n_y, n_features), "
+                f"got {inputs.shape}"
+            )
+        graph = build_dgl_graph(nx_tile=inputs.shape[2])
+        reshaped = inputs.reshape(
+            inputs.shape[0],
+            inputs.shape[1] * inputs.shape[2] * inputs.shape[3],
+            inputs.shape[4],
+        ).transpose(0, 1)
+        convolved = self.graph_op(graph, reshaped).transpose(1, 0)
+        return convolved.reshape(
+            inputs.shape[0], inputs.shape[1], inputs.shape[2], inputs.shape[3], -1
+        )
+
+
+class GraphNetwork(nn.Module):
+    def __init__(self, config, n_features_in: int, n_features_out: int):
+        super(GraphNetwork, self).__init__()
+        self.conv1 = CubedSphereGraphOperation(
```

Comment:
For the model implemented here, it would probably make more sense to wrap the entire GraphNetwork in a CubedSphereGraphOperation, since it's only doing graph operations. However this model is outdated, we plan to update it to the U-net implementation which will require doing a pooling or upsampling operation in cube-space between each convolutional block. This will be a better starting point for that update.

The operations it does are also very cheap - reshaping an index twice, neither of which copy memory, and retrieving a graph from LRU cache. Performance-wise, it shouldn't really matter which one of these we do.

Reply:
Cool, thanks for the information.